### PR TITLE
Added support from OID-side port association

### PR DESCRIPTION
**Pull request by Attil**

### CODE REVIEW

File: windows/vr_assoc.c [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	{
+		hash ^= char_map[(*str++) % 256] * 5;
+	}
+
+	return hash;
+}
+
+struct vr_assoc* vr_get_assoc(NDIS_IF_COUNTED_STRING string)
+{
+	struct vr_assoc** field = map + vr_hash_nic(string);
```
Comment (m-kostrzewa) [inline windows/vr_assoc.c]: I don't get why + instead of []? This may be confusing I think.

File: windows/vrouter_mod.c [hunk 2]
```diff
@@ -9,8 +10,49 @@ PWCHAR SxExtUniqueName = L"{56553588-1538-4BE6-B8E0-CB46402DC205}";
...
+	'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P',
+	'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X',
+	'Y', 'Z', 'a', 'b', 'c', 'd', 'e', 'f',
+	'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n',
+	'o', 'p', 'q', 'r', 's', 't', 'u', 'v',
+	'w', 'x', 'y', 'z', '0', '1', '2', '3',
+	'4', '5', '6', '7', '8', '9', '+', '/' };
+static int mod_table[] = { 0, 2, 1 };
+
+char *base64_encode(const unsigned char *data,
```
Comment (m-kostrzewa) [inline windows/vrouter_mod.c]: unnecessary line split

File: windows/vrouter_mod.c [hunk 3]
```diff
@@ -93,10 +141,40 @@ SxExtCreatePort(
...
 	UNREFERENCED_PARAMETER(ExtensionContext);
-	UNREFERENCED_PARAMETER(Port);
+
+	DbgPrint("SxExtCreatePort\r\n");
+	DbgPrint("PortFriendlyName: %S, PortName: %S, PortId: %u, PortState: %u, PortType: %u\r\n", Port->PortFriendlyName.String, Port->PortName.String, Port->PortId, Port->PortState, Port->PortType);
+	NDIS_IF_COUNTED_STRING id = vr_get_name_from_friendly_name(Port->PortFriendlyName);
+	DbgPrint("Resolved name: %S\r\n", id.String);;
+	vr_set_assoc_oid(id, Port->PortId);
+
+	PNDIS_SWITCH_PORT_PROPERTY_ENUM_PARAMETERS portPropertyParameters = NULL;
```
Comment (m-kostrzewa) [inline windows/vrouter_mod.c]: what's up with all this VLAN handling?

File: windows/vrouter_mod.c [hunk 4]
```diff
@@ -93,10 +141,40 @@ SxExtCreatePort(
...
 )
 {
-	DbgPrint("SxExtCreatePort\r\n");
 	UNREFERENCED_PARAMETER(Switch);
 	UNREFERENCED_PARAMETER(ExtensionContext);
-	UNREFERENCED_PARAMETER(Port);
+
+	DbgPrint("SxExtCreatePort\r\n");
+	DbgPrint("PortFriendlyName: %S, PortName: %S, PortId: %u, PortState: %u, PortType: %u\r\n", Port->PortFriendlyName.String, Port->PortName.String, Port->PortId, Port->PortState, Port->PortType);
+	NDIS_IF_COUNTED_STRING id = vr_get_name_from_friendly_name(Port->PortFriendlyName);
```
Comment (m-kostrzewa) [inline windows/vrouter_mod.c]: I think we should either stick to 'id' or 'name' here.

File: windows/vrouter_mod.c [hunk 5]
```diff
@@ -472,24 +553,111 @@ SxExtStartNetBufferListsIngress(
...
 
 	sendCompleteFlags |= (dispatch) ? NDIS_SEND_COMPLETE_FLAGS_DISPATCH_LEVEL : 0;
 	SendFlags |= NDIS_SEND_FLAGS_SWITCH_DESTINATION_GROUP;
 
+	NdisAcquireRWLockRead(lock, &lockState, dispatch);
+
 	if (sameSource)
 	{
+		sendCompleteFlags |= NDIS_SEND_COMPLETE_FLAGS_SWITCH_SINGLE_SOURCE;
 		DbgPrint("--------Hehe, sejmsors!!!!!\r\n");
```
Comment (m-kostrzewa) [inline windows/vrouter_mod.c]: Could you improve these debug messages?

File: windows/vrouter_mod.c [hunk 6]
```diff
@@ -472,24 +553,111 @@ SxExtStartNetBufferListsIngress(
...
+	}
+
+	if (dropNbl != NULL)
+	{
+		SxLibCompleteNetBufferListsIngress(Switch,
+			dropNbl,
+			sendCompleteFlags);
+	}
+
+	NdisReleaseRWLock(lock, &lockState);
```
Comment (m-kostrzewa) [inline windows/vrouter_mod.c]: Is this NBL handling necessary for this PR? Maybe it should be moved to another PR?

File: windows/vr_assoc.c (hunk 1 above)
Comment (Attil) [inline windows/vr_assoc.c]: Using [] also dereferences one level, a pointer to the pointer is used later.

File: include/vr_windows.h [hunk 7]
```diff
@@ -0,0 +1,24 @@
+#pragma once
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define VR_OID_SOURCE	0x00000001
+#define VR_AGENT_SOURCE	0x00000002
```
Comment (Attil) [inline include/vr_windows.h]: No, its a bitmask. The next one would be 0x00000004

**Comment by Attil** [comment]
I applied some of the suggestions and split this PR into two smaller PRs.

File: vRouter.vcxproj [hunk 8]
```diff
@@ -202,7 +202,7 @@
       <WarningLevel>Level4</WarningLevel>
       <TreatWarningAsError>true</TreatWarningAsError>
       <AdditionalIncludeDirectories>..;.;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
-      <PreProcessorDefinitions>%(PreProcessorDefinitions);NDIS_WDM=1;NDIS630=1</PreProcessorDefinitions>
+      <PreProcessorDefinitions>%(PreProcessorDefinitions);NDIS_WDM=1;NDIS640=1</PreProcessorDefinitions>
```
Comment (m-kostrzewa) [inline vRouter.vcxproj]: Pawel says 640 is fine

File: windows/vr_assoc.c [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+	while (*field != NULL)
+	{
+		if ((*field)->string.Length == interface_name.Length && wcsncmp((*field)->string.String, interface_name.String, interface_name.Length) == 0)
+			break;
+		field = &((*field)->next);
+	}
+
+	if (*field == NULL)
+	{
+		*field = ExAllocatePoolWithTag(NonPagedPoolNx, sizeof(struct vr_assoc), SxExtAllocationTag);
```
Comment (michal-clapinski) [inline windows/vr_assoc.c]: Didn't you write some functions for allocating memory as a part of "basic Windows callbacks"? Shouldn't you use them instead of plain "ExAllocatePoolWithTag"?

File: windows/vr_assoc.c [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+void vr_assoc_destroy()
+{
+	struct vr_assoc* next_element;
+	struct vr_assoc* element;
+
+	for (int i = 0; i < MAP_SIZE; i++)
+	{
+		next_element = map[i];
```
Comment (michal-clapinski) [inline windows/vr_assoc.c]: Looks useless to me.

File: windows/vrouter_mod.c [hunk 11]
```diff
@@ -220,10 +226,11 @@ SxExtSaveNic(
 	UNREFERENCED_PARAMETER(Switch);
 	UNREFERENCED_PARAMETER(ExtensionContext);
 	UNREFERENCED_PARAMETER(SaveState);
-	UNREFERENCED_PARAMETER(BytesWritten);
-	UNREFERENCED_PARAMETER(BytesNeeded);
 
-	return 0;
+	*BytesWritten = 0;
+	*BytesNeeded = 0;
```
Comment (michal-clapinski) [inline windows/vrouter_mod.c]: what does it do?

File: windows/vr_assoc.c (hunk 9 above)
Comment (Attil) [inline windows/vr_assoc.c]: It'd be possible to use them, but right now that part requires some more dp-core porting. And everything in windows/ directory is windows-specific so there's no need to add an abstraction layer there.

Comment (michal-clapinski) [inline windows/vr_assoc.c]: I think using them increases readability of code. "ExAllocatePoolWithTag(NonPagedPoolNx..." isn't the most pleasant thing to see.

File: windows/vrouter_mod.c (hunk 11 above)
Comment (Attil) [inline windows/vrouter_mod.c]: BytesNeeded tells SxBase how many bytes should it allocate.

BytesWritten tells SxBase how many bytes we've saved. 0 means we didn't save anything and it is treated differently to other cases inside.

**Comment by Attil** [comment]
Bump

File: include/vr_windows.h [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+};
+
+NDIS_IF_COUNTED_STRING vr_get_name_from_friendly_name(NDIS_IF_COUNTED_STRING friendly);
+int vr_hash_nic(NDIS_IF_COUNTED_STRING string);
+struct vr_assoc* vr_get_assoc(NDIS_IF_COUNTED_STRING string);
+void vr_set_assoc_oid(NDIS_IF_COUNTED_STRING name, NDIS_SWITCH_PORT_ID port_id);
+
+#ifdef __cplusplus
+}
+#endif
```
Comment (michal-clapinski) [inline include/vr_windows.h]: enter

File: windows/precomp.h [hunk 13]
```diff
@@ -8,6 +8,5 @@
 #include "SxBase\SxApi.h"
 #include "SxBase\SxLibrary.h"
 
-//#include "vrouter.h"
-
-ULONG  SxExtAllocationTag = 'RVCO';
\ No newline at end of file
+#include "vr_windows.h"
+//#include "vrouter.h"
```
Comment (michal-clapinski) [inline windows/precomp.h]: hm

File: windows/vr_assoc.c [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+	{
+		element = map[i];
+		while (element)
+		{
+			next_element = element->next;
+			ExFreePoolWithTag(element, SxExtAllocationTag);
+			element = next_element;
+		}
+	}
+}
```
Comment (michal-clapinski) [inline windows/vr_assoc.c]: here

**Comment by michal-clapinski** [comment]
http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206

File: windows/vrouter_mod.c (hunk 11 above)
Comment (m-kostrzewa) [inline windows/vrouter_mod.c]: Required by extension API. AFAIK it has something to do with persistence

File: windows/vr_assoc.c (hunk 14 above)
Comment (m-kostrzewa) [inline windows/vr_assoc.c]: I think we should decide on a c++ linter
